ios/browser/api/qr_code/qr_code_generator.mm

```diff
@@ -0,0 +1,389 @@
+
+#include "brave/ios/browser/api/qr_code/qr_code_generator.h"
+
+#include "base/bind.h"
+#include "base/callback.h"
+#include "base/strings/sys_string_conversions.h"
+
+#include "mojo/public/cpp/bindings/pending_receiver.h"
+#include "mojo/public/cpp/bindings/receiver.h"
+#include "third_party/skia/include/core/SkCanvas.h"
+
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "base/strings/string_util.h"
+#include "components/qr_code_generator/dino_image.h"
+#include "components/qr_code_generator/qr_code_generator.h"
+#include "ui/gfx/canvas.h"
+#include "ui/gfx/geometry/size.h"
+#include "ui/gfx/geometry/skia_conversions.h"
+#include "ui/gfx/image/image.h"
+#include "ui/gfx/image/image_skia.h"
+
+#if !defined(__has_feature) || !__has_feature(objc_arc)
+#error "This file requires ARC support."
+#endif
+
+// MARK: - Implementation
+
+namespace qrcode_generator {
+
+// Allow each element to render as this many pixels.
+static const int kModuleSizePixels = 10;
+
+// Allow each dino tile to render as this many pixels.
+static const int kDinoTileSizePixels = 4;
+
+// Size of a QR locator, in modules.
+static const int kLocatorSizeModules = 7;
+}  // namespace qrcode_generator
+
+// MARK: - BraveQRCodeGeneratorOptions
+@interface BraveQRCodeGeneratorOptions ()
+@property(nonatomic, readonly) NSString* data;
+@property(nonatomic, readonly) bool shouldRender;
+@property(nonatomic, readonly) bool renderDino;
+@property(nonatomic, readonly)
+    BraveQRCodeGeneratorModuleStyle renderModuleStyle;
+@property(nonatomic, readonly)
+    BraveQRCodeGeneratorLocatorStyle renderLocatorStyle;
+@end
+
+@implementation BraveQRCodeGeneratorOptions
+- (instancetype)initWithData:(NSString*)data
+                shouldRender:(bool)shouldRender
+          renderLogoInCenter:(bool)renderLogoInCenter
+           renderModuleStyle:(BraveQRCodeGeneratorModuleStyle)renderModuleStyle
+          renderLocatorStyle:
+              (BraveQRCodeGeneratorLocatorStyle)renderLocatorStyle {
+  if ((self = [super init])) {
+    _data = data;
+    _shouldRender = shouldRender;
+    _renderDino = renderLogoInCenter;
+    _renderModuleStyle = renderModuleStyle;
+    _renderLocatorStyle = renderLocatorStyle;
+  }
+  return self;
+}
+@end
+
+// MARK: - BraveQRCodeGeneratorResult
+@interface BraveQRCodeGeneratorResult ()
+@property(nonatomic) BraveQRCodeGeneratorError errorCode;
+@property(nonatomic) UIImage* image;
+@property(nonatomic) NSData* data;
+@property(nonatomic) CGSize dataSize;
++ (instancetype)newInstance;
+@end
+
+@implementation BraveQRCodeGeneratorResult
++ (instancetype)newInstance {
+  return [[BraveQRCodeGeneratorResult alloc] init];
+}
+@end
+
+// MARK: - BraveQRCodeGenerator
+
+@interface BraveQRCodeGenerator () {
+  SkBitmap dino_bitmap_;
+}
+@end
+
+@implementation BraveQRCodeGenerator
+- (instancetype)init {
+  if ((self = [super init])) {
+    [self initializeDinoBitmap];
+  }
+  return self;
+}
+
+- (void)initializeDinoBitmap {
```

Comment:
Weird ask, but can we possibly move all of the stuff copied from https://source.chromium.org/chromium/chromium/src/+/main:chrome/services/qrcode_generator/qrcode_generator_service_impl.cc outside of `api` and move it to to a `ios/browser/qr_codegenerator`? And then in that new set of files we can just straight up add a big comment at the top of each saying its basically copied from Chromium but we can't use them due to imports to `//chrome`? Then the API in this file becomes much cleaner, essentially just a wrapper around the copied C++

Reply:
Done.